sklearn/metrics/_pairwise_distances_reduction/_argkminlabels.pyx

```diff
@@ -0,0 +1,214 @@
+from cython cimport floating, integral
+from cython.parallel cimport parallel, prange
+from libcpp.map cimport map as cmap
+from libc.stdlib cimport free
+
+cimport numpy as cnp
+
+cnp.import_array()
+
+from ._argkmin cimport PairwiseDistancesArgKmin64
+from ._datasets_pair cimport DatasetsPair, DenseDenseDatasetsPair
+from ...utils._typedefs cimport ITYPE_t, DTYPE_t
+from ...utils._typedefs import ITYPE, DTYPE
+from ...utils._sorting cimport simultaneous_sort
+import numpy as np
+
+from sklearn.utils.fixes import threadpool_limits
+
+
+cpdef enum WeightingStrategy:
+    uniform = 0
+    distance = 1
+    other = 2
+
+cdef inline ITYPE_t weighted_histogram_mode(ITYPE_t sample_index, ITYPE_t k,
+                                    ITYPE_t* indices, const ITYPE_t[:] labels,
+                                    cmap[ITYPE_t, ITYPE_t] labels_to_index,
+                                    DTYPE_t* label_weights, DTYPE_t* distances,
+                                    WeightingStrategy weight_type) nogil:
```

Comment:
If we keep `weighted_histogram_mode` as is, it might be better to use `const` qualification.

Also nit: black-formatting might make it more readable. What do you think?

```suggestion
cdef inline ITYPE_t weighted_histogram_mode(
    ITYPE_t sample_index,
    ITYPE_t k,
    const ITYPE_t* indices,
    const ITYPE_t[:] labels,
    const cmap[ITYPE_t, ITYPE_t] labels_to_index,
    DTYPE_t* label_weights,
    const DTYPE_t* distances,
    WeightingStrategy weight_type,
) nogil:
```